src/meshapi/util/uisp_import/sync_handlers.py

```diff
@@ -0,0 +1,297 @@
+import datetime
+import logging
+import re
+from typing import List
+
+from django.db import transaction
+from django.db.models import Q
+
+from meshapi.models import LOS, Device, Link, Node, Sector
+from meshapi.serializers import DeviceSerializer, LinkSerializer
+from meshapi.types.uisp_api.data_links import DataLink as UISPDataLink
+from meshapi.types.uisp_api.devices import Device as UISPDevice
+from meshapi.util.admin_notifications import notify_administrators_of_data_issue
+from meshapi.util.uisp_import.constants import (
+    DEFAULT_SECTOR_AZIMUTH,
+    DEFAULT_SECTOR_RADIUS,
+    DEFAULT_SECTOR_WIDTH,
+    DEVICE_MODEL_TO_BEAM_WIDTH,
+    DEVICE_NAME_NETWORK_NUMBER_SUBSTITUTIONS,
+    EXCLUDED_UISP_DEVICE_CATEGORIES,
+    NETWORK_NUMBER_REGEX_FOR_DEVICE_NAME,
+)
+from meshapi.util.uisp_import.fetch_uisp import get_uisp_session
+from meshapi.util.uisp_import.update_objects import update_device_from_uisp_data, update_link_from_uisp_data
+from meshapi.util.uisp_import.utils import (
+    downclass_device,
+    get_building_from_network_number,
+    get_link_type,
+    guess_compass_heading_from_device_name,
+    notify_admins_of_changes,
+    parse_uisp_datetime,
+)
+
+
+def import_and_sync_uisp_devices(uisp_devices: List[UISPDevice]) -> None:
+    for uisp_device in uisp_devices:
+        uisp_uuid = uisp_device["identification"]["id"]
+        uisp_category = uisp_device["identification"]["category"]
+        uisp_name = uisp_device["identification"]["name"]
+
+        if uisp_category in EXCLUDED_UISP_DEVICE_CATEGORIES:
+            logging.debug(
+                f"During UISP device import, {uisp_name} "
+                f'(UISP ID {uisp_uuid}) was skipped because the "{uisp_category}" has been '
+                "excluded in meshapi.util.uisp_import.constants"
+            )
+            continue
+
+        modified_name = uisp_device["identification"]["name"]
+        for find, replace in DEVICE_NAME_NETWORK_NUMBER_SUBSTITUTIONS.items():
+            modified_name = modified_name.replace(find, replace)
+
+        network_number_matches = re.findall(NETWORK_NUMBER_REGEX_FOR_DEVICE_NAME, modified_name)
```

Comment:
Oh boy, a regex. I see that you've got tests for this function, so I guess I'll let it slide. Can you show me some examples of what the regex is supposed to match?

Reply:
It extracts the the NN from the device name, which is the only way I can think of to do the mapping. This is a bit tricky for p2p radios, but fortunately the convention is that the NN the device is located at is the first number, and the second number is the remote side. Some examples:
```
nycmesh-227-north -> 227
nycmesh-5916-lbe-3606 -> 5916
nycmesh-lbe-238 -> 238
``